addon/-private/validations-for.js

```diff
@@ -0,0 +1,32 @@
+const VALIDATION_MAP = new WeakMap();
+
+/**
+ * Recursively get all validations for the given class, and all parent classes
+ */
+export function getValidationsFor(klass) {
+  const klassValidations = VALIDATION_MAP.get(klass) || {};
+  const prototype = Object.getPrototypeOf(klass);
+  let prototypeValidations = {};
+
+  if (prototype) {
+    prototypeValidations = getValidationsFor(prototype);
+  }
+
+  // Note: order is important to ensure that a subclass cannot override
+  // the type of an argument
+  return { ...klassValidations, ...prototypeValidations };
```

Comment:
Hmm... that's interesting. We'd need to introduce a kind of hierarchy between types, right?

- `'object' => shapeOf({ what: 'ever' })`
- `'array' => arrayOf('string')`
- `'any' => literally anything else`

Reply:
The way it originally worked was that we validated against all validators in the inheritance tree, if I remember correctly. So, in the above case, you would validate against _both_ `any` and `string`. This would prevent incoherent validations:

```js
class Foo {
  @argument('number') foo;
}

class Bar extends Foo {
  @argument('string') foo;
}
```

Would always fail one or the other, so users would never be able to pass in `foo`. Hard to add a good error message here though, so maybe this is too complicated and we should just either allow children to override parents, or keep the current behavior.